nixos/no-x-libs: add more replacements

Artturin opened the pull request:
all-packages: change -nox overrides to callPackage's so there is no infinite recursion when replacing the package with a overlay

###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

RaitoBezarius commented:
@Artturin Is there an easy way to measure the delta in the closure size?

Artturin commented:
> @Artturin Is there an easy way to measure the delta in the closure size?

this doesn't change the closure size much (atleast for a minimal-ish system assuming i did the comparison correctly) but instead reduces the dependency tree size ( reducing compilation times)

```
{
  inputs = {
    nixpkgs.url = "github:NixOS/nixpkgs/refs/pull/204287/head";
    # commit before the above pr
    nixpkgs-before.url = "github:NixOS/nixpkgs/7bfceb02093f2c7ea0ea7e353c99b26e9a4022fc";
  };

  outputs = inputs:
    let
      pkgs = import inputs.nixpkgs { system = "x86_64-linux"; };
      pkgs-before = import inputs.nixpkgs-before { system = "x86_64-linux"; };

      node = pkgs.nixos ({ modulesPath, ... }: {
        fileSystems."/".device = "ignore-root-device";
        boot.loader.grub.enable = false;
        imports = [
          (modulesPath + "/profiles/minimal.nix")
        ];
        networking.networkmanager = {
          enable = true;
          plugins = pkgs.lib.mkForce [];
        };
      });
      node-before = pkgs-before.nixos ({ modulesPath, ... }: {
        fileSystems."/".device = "ignore-root-device";
        boot.loader.grub.enable = false;
        imports = [
          (modulesPath + "/profiles/minimal.nix")
        ];
        networking.networkmanager = {
          enable = true;
          plugins = pkgs.lib.mkForce [];
        };
        nixpkgs.overlays = [
          (_: super: {
            stoken = super.stoken.override { withGTK3 = false; };
          })
        ];
      });

      closureInfo = pkgs.closureInfo { rootPaths = [ node.config.system.build.toplevel ]; };
      closureInfoBefore = pkgs.closureInfo { rootPaths = [ node-before.config.system.build.toplevel ]; };
    in
    {
      packages.x86_64-linux.default = pkgs.runCommand "compare"
        {
          inherit closureInfo closureInfoBefore;
          nativeBuildInputs = [ pkgs.nodePackages.insect ];
        } ''
        sizeBytes=$(cat $closureInfo/total-nar-size)
        sizeBeforeBytes=$(cat $closureInfoBefore/total-nar-size)
        insect "$((sizeBeforeBytes - sizeBytes)) bytes -> megabytes"
        touch $out
      '';
      packages.x86_64-linux.has-libX = closureInfoBefore;
      packages.x86_64-linux.no-libX = closureInfo;
    };
}
```

```
2167
$ nix path-info -rsSh ".#packages.x86_64-linux.no-libX" --derivation | wc -l
1866
```

SuperSandro2000 commented on nixos/modules/config/no-x-libs.nix:
```diff
@@ -42,6 +42,14 @@ with lib;
+      # translateManpages -> perlPackages.po4a -> texlive-combined-basic -> texlive-core-big -> libX11
+      util-linux = super.util-linux.override { translateManpages = false; };
+      # dep of graphviz, libXpm is optional for Xpm support
+      gd = super.gd.override { libXpm = null; };
```
Can we please add an option instead of overwriting packages to null? This would for example break if we access libXpm.version.

SuperSandro2000 replied:
https://github.com/NixOS/nixpkgs/pull/204426

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -7512,7 +7512,8 @@ with pkgs;
   };
 
-  graphviz-nox = graphviz.override {
+  graphviz-nox = callPackage ../tools/graphics/graphviz {
+    inherit (darwin.apple_sdk.frameworks) ApplicationServices;
```
Please change those back to override. Otherwise we need to maintain inherits at two places.

Artturin replied:
That will cause infinite recursion

SuperSandro2000 replied:
Can you research where that happens? I would assume that if we are using less libraries then this should not happen?

Artturin replied:
The overlay replaces graphviz with graphviz-nox  and graphviz-nox is graphviz overridden 

graphviz -> graphviz-nox --> graphviz

Artturin replied:
@roberth also said the same thing 
![image](https://user-images.githubusercontent.com/56650223/205523876-5ee177fa-d717-427f-82dd-fc4906ded5b4.png)

SuperSandro2000 replied:
Oh, okay. Then we should add a very short comment so that we still know this later.

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -20151,7 +20152,8 @@ with pkgs;
-  libdevil-nox = libdevil.override {
+  libdevil-nox = callPackage ../development/libraries/libdevil {
+    inherit (darwin.apple_sdk.frameworks) OpenGL;
     libX11 = null;
     libGL = null;
```
Let me create a PR that adds a proper option for this.

SuperSandro2000 replied:
https://github.com/NixOS/nixpkgs/pull/204425

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -20151,7 +20152,8 @@ with pkgs;
   };
 
-  libdevil-nox = libdevil.override {
+  libdevil-nox = callPackage ../development/libraries/libdevil {
+    inherit (darwin.apple_sdk.frameworks) OpenGL;
     libX11 = null;
     libGL = null;
   };
```
```suggestion
  libdevil-nox = libdevil.override { withXorg = false; };
```

SuperSandro2000 commented:
FYI you motivated to try this again especially for my simd branch and I created the following PRs:
- #204545
- #204564
- #204573

roberth commented:
Related
 - #204977

> I'm not doing this through the NixOS no-x-libs overlay, because this should improve many packages that have no business having graphical libraries in their build closure, saving many rebuilds, regardless of whether the user wants to optimize for a non-graphical system configuration.

Artturin committed:
nixos/no-x-libs: add more replacements

all-packages: change -nox overrides to callPackage's so there is no
infinite recursion when replacing the package with a overlay

Artturin commented:
wait until https://nixpk.gs/pr-tracker.html?pr=204426 is in master

SuperSandro2000 commented:
Why not merge this into staging?

nagy commented on nixos/modules/config/no-x-libs.nix:
```diff
@@ -45,9 +49,13 @@ with lib;
       qrencode = super.qrencode.overrideAttrs (_: { doCheck = false; });
+      stoken = super.stoken.override { withGTK3 = false; };
+      # translateManpages -> perlPackages.po4a -> texlive-combined-basic -> texlive-core-big -> libX11
+      util-linux = super.util-linux.override { translateManpages = false; };
```
`po4a` is only a build-time dependency. Is an override for this really needed?

SuperSandro2000 replied:
Yes, noXLibs removes X already at build time and this breaks compiling some packages.

SuperSandro2000 commented:
Needed to create a new PR because force pushing that one almost deadlocked my computer. #211428

Izorkin commented:
This PR causes this error:
```
error: infinite recursion encountered

       at /home/lafiel/works/src-nix/nixpkgs/pkgs/top-level/all-packages.nix:20695:3:

        20694|
        20695|   libdevil-nox = libdevil.override {
             |   ^
        20696|     inherit (darwin.apple_sdk.frameworks) OpenGL;
(use '--show-trace' to show detailed location information)
```

Artturin commented on nixos/modules/config/no-x-libs.nix:
```diff
@@ -32,13 +32,17 @@ with lib;
-      gpsd = super.gpsd.override { guiSupport = false; };
       imagemagick = super.imagemagick.override { libX11Support = false; libXtSupport = false; };
       imagemagickBig = super.imagemagickBig.override { libX11Support = false; libXtSupport = false; };
+      libdevil = super.libdevil-nox;
```
libdevil -> libdevil-nox -> libdevil -> libdevil-nox ...

Artturin replied:
changing this back to

```
  libdevil-nox = callPackage ../development/libraries/libdevil {
    inherit (darwin.apple_sdk.frameworks) OpenGL;
    withXorg = false;
  };
```
will work

Izorkin replied:
Thanks, working!

SuperSandro2000 replied:
Ahh, sorry, messed this up when rebasing.

SuperSandro2000 replied:
#211843

Artturin replied:
Thanks @SuperSandro2000
